aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,312 @@
+  :language: c#
+  :emphasize-lines: 11-12
+
+Filters and Attributes
+^^^^^^^^^^^^^^^^^^^^^^
+It's often convenient to implement filter interfaces as *Attributes*. These attributes are then used to apply the filter to certain controllers or action methods. The framework includes built-in attribute-based filters that you can subclass and customize. For example, the following filter inherits from `ResultFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ResultFilterAttribute/index.html>`_, and overrides its ``OnResultExecuting`` method to add a header to the response.
+
+.. _add-header-attribute:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/AddHeaderAttribute.cs
+  :language: c#
+  :emphasize-lines: 5,16
+
+Attributes allow filters to accept arguments, as shown in the example above. You would add this attribute to a controller or action method and specify the name and value of the HTTP header you wished to add to the response:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Controllers/SampleController.cs
+  :language: c#
+  :emphasize-lines: 1
+  :lines: 6-12,25
+  :dedent: 4
+
+The result of the ``Index`` action is shown below - the response headers are displayed on the bottom right.
+
+.. image:: filters/_static/add-header.png
+
+The available built-in attribute-based filters are:
+	- `ActionFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ActionFilterAttribute/index.html>`_
+	- `AuthorizationFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/AuthorizationFilterAttribute/index.html>`_
+	- `ExceptionFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ExceptionFilterAttribute/index.html>`_
+	- `ResultFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ResultFilterAttribute/index.html>`_
+
+Dependency Injection
+^^^^^^^^^^^^^^^^^^^^
+Filters are not directly returned from :doc:`dependency injection </fundamentals/dependency-injection>` (DI). Filters that are also attributes, and which are added directly to controller classes or action methods cannot have constructor dependencies provided by DI - their constructor properties must be specified when they are declared. This is a limitation of how attributes work. However, if your filters have dependencies you need to access from DI, there are several supported approaches. You can apply your filter using either the `ServiceFilter <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/ServiceFilterAttribute/index.html>`_ or `TypeFilter <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/TypeFilterAttribute/index.html>`_ attribute, or you can implement IFilterFactory on your attribute, or you can access the services you require from the context parameter provided to your filter's On\ *Stage* methods.
```

Comment:
boarder line too many or and too long. See what it looks like as

 You can apply your filter using:

- the ServiceFilter or TypeFilter attribute
- Implement IFilterFactory on your attribute,
- ou can access the services you require from the context parameter provided to your filter’s OnStage methods.

Needs some word smithing.